tool/src/org/antlr/v4/semantics/SymbolChecks.java

```diff
@@ -341,42 +325,156 @@ public void checkForModeConflicts(Grammar g) {
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * {@return} list of simple string literals for rule {@param rule}
+	 */
+	private List<String> getSingleTokenValues(Rule rule)
+	{
+		List<String> values = new ArrayList<>();
+		for (Alternative alt : rule.alt) {
+			if (alt != null) {
+				// select first alt if token has a command
+				Tree rootNode = alt.ast.getChildCount() == 2 &&
+						alt.ast.getChild(0) instanceof AltAST && alt.ast.getChild(1) instanceof GrammarAST
+						? alt.ast.getChild(0)
+						: alt.ast;
+
+				if (rootNode.getTokenStartIndex() == -1) {
+					continue; // ignore autogenerated tokens from combined grammars that start with T__
+				}
+
+				// Ignore alt if contains not only string literals (repetition, optional)
+				boolean ignore = false;
+				StringBuilder currentValue = new StringBuilder();
+				for (int i = 0; i < rootNode.getChildCount(); i++) {
+					Tree child = rootNode.getChild(i);
+					if (!(child instanceof TerminalAST)) {
+						ignore = true;
+						break;
+					}
+
+					TerminalAST terminalAST = (TerminalAST)child;
+					if (terminalAST.token.getType() != ANTLRLexer.STRING_LITERAL) {
+						ignore = true;
+						break;
+					}
+					else {
+						String text = terminalAST.token.getText();
+						currentValue.append(text.substring(1, text.length() - 1));
+					}
+				}
+
+				if (!ignore) {
+					values.add(currentValue.toString());
+				}
+			}
+		}
+		return values;
+	}
+
+	/**
+	 * For same rule compare values from next index:
+	 * TOKEN_WITH_SAME_VALUES: 'asdf' | 'asdf';
+	 * For different rules compare from start value:
+	 * TOKEN1: 'asdf';
+	 * TOKEN2: 'asdf';
+	 */
+	private void checkForOverlap(Grammar g, Rule rule1, Rule rule2, List<String> firstTokenStringValues, List<String> secondTokenStringValues) {
+		for (int i = 0; i < firstTokenStringValues.size(); i++) {
+			int secondTokenInd = rule1 == rule2 ? i + 1 : 0;
+			String str1 = firstTokenStringValues.get(i);
+			for (int j = secondTokenInd; j < secondTokenStringValues.size(); j++) {
+				String str2 = secondTokenStringValues.get(j);
+				if (str1.equals(str2)) {
+					errMgr.grammarError(ErrorType.TOKEN_UNREACHABLE, g.fileName,
+							((GrammarAST) rule2.ast.getChild(0)).token, rule2.name, str2, rule1.name);
+				}
+			}
+		}
+	}
+
+	// CAN ONLY CALL THE TWO NEXT METHODS AFTER GRAMMAR HAS RULE DEFS (see semanticpipeline)
 	public void checkRuleArgs(Grammar g, List<GrammarAST> rulerefs) {
-		if ( rulerefs==null ) return;
+		if (rulerefs == null) return;
 		for (GrammarAST ref : rulerefs) {
 			String ruleName = ref.getText();
 			Rule r = g.getRule(ruleName);
-			GrammarAST arg = (GrammarAST)ref.getFirstChildWithType(ANTLRParser.ARG_ACTION);
-			if ( arg!=null && (r==null || r.args==null) ) {
+			GrammarAST arg = (GrammarAST) ref.getFirstChildWithType(ANTLRParser.ARG_ACTION);
+			if (arg != null && (r == null || r.args == null)) {
```

Comment:
:memo: Spacing here was generally unspecified (@parrt uses the previous spacing, while I use the updated spacing). Generally accepted practice became not changing it unless you were rewriting the line.